pkg/antctl/raw/upgrade/apistorage/command_test.go

```diff
@@ -0,0 +1,135 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apistorage
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	apiextinstall "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var (
+	antreaCRDAlfa       = createMockCRD("Alfa", "AlfaList", "alfas", "alfa", "crd.antrea.io", map[string]bool{"v1alpha1": false, "v1beta1": true})
+	antreaCRDBravo      = createMockCRD("Bravo", "BravoList", "bravos", "bravo", "crd.antrea.io", map[string]bool{"v1beta1": true})
+	nonAntreaCRDCharlie = createMockCRD("Charlie", "CharlieList", "charlies", "charlie", "crd.misc.io", map[string]bool{"v1beta1": true})
+	alfa1               = createMockCRDObject("crd.antrea.io/v1beta1", "Alfa", "alfa1")
+	alfa2               = createMockCRDObject("crd.antrea.io/v1beta1", "Alfa", "alfa2")
+	bravo1              = createMockCRDObject("crd.antrea.io/v1beta1", "Bravo", "bravo1")
+)
+
+func createMockCRD(kind, listKind, plural, singular, group string, versions map[string]bool) *apiextv1.CustomResourceDefinition {
+	crd := &apiextv1.CustomResourceDefinition{}
+	var crdVersions []apiextv1.CustomResourceDefinitionVersion
+	var crdStoredVersions []string
+	for version, served := range versions {
+		crdVersions = append(crdVersions, apiextv1.CustomResourceDefinitionVersion{
+			Name:    version,
+			Served:  served,
+			Storage: served,
+		})
+		crdStoredVersions = append(crdStoredVersions, version)
+	}
+
+	crd.SetName(fmt.Sprintf("%s.%s", plural, group))
+	crd.Spec = apiextv1.CustomResourceDefinitionSpec{
+		Group: group,
+		Names: apiextv1.CustomResourceDefinitionNames{
+			Kind:     kind,
+			ListKind: listKind,
+			Plural:   plural,
+			Singular: singular,
+		},
+		Versions: crdVersions,
+	}
+	crd.Status.StoredVersions = crdStoredVersions
+	return crd
+}
+
+func createMockCRDObject(apiVersion, kind, name string) *unstructured.Unstructured {
+	obj := &unstructured.Unstructured{}
+	obj.SetAPIVersion(apiVersion)
+	obj.SetNamespace("default")
+	obj.SetName(name)
+	obj.SetKind(kind)
+	return obj
+}
+
+func TestUpgradeCRDs(t *testing.T) {
+
+	tests := []struct {
+		name           string
+		crdsToUpgrade  []string
+		originalCRDs   []client.Object
+		expectedOutput string
+	}{
+		{
+			name:         "Upgrade all CRDs",
+			originalCRDs: []client.Object{antreaCRDAlfa, antreaCRDBravo, nonAntreaCRDCharlie, alfa1, alfa2, bravo1},
+			expectedOutput: `No need to upgrade CRD "bravos.crd.antrea.io" since it only has one version.
+Upgrading 2 CRD "alfas.crd.antrea.io" objects.
+Successfully upgraded 2 CRD alfas.crd.antrea.io objects.
```

Comment:
ditto, better to have double quotes around `alfas.crd.antrea.io` like above output.